R/detectSpatialContext.R

```diff
@@ -0,0 +1,109 @@
+#'  \code{colData(object)}. Defaults to "spatial_context".
+#'
+#' @return returns an object of \code{class(object)} containing a new column 
+#' entry to \code{colData(object)[[name]]}
+#' 
+#' @examples 
+#' set.seed(22)
+#' library(cytomapper)
+#' data(pancreasSCE)
+#'
+#' ## 1. Cellular neighborhood (CN)
+#' sce <- buildSpatialGraph(pancreasSCE, img_id = "ImageNb",
+#'                         type = "knn",
+#'                         name = "knn_cn_graph",
+#'                         k = 5)
+#'
+#' sce <- aggregateNeighbors(sce, colPairName = "knn_cn_graph",
+#'                          aggregate_by = "metadata",
+#'                          count_by = "CellType",
+#'                          name = "aggregatedCellTypes")
+#'
+#' cur_cluster <- kmeans(sce$aggregatedCellTypes, centers = 3)
+#' sce$cellular_neighborhood <- factor(cur_cluster$cluster)
+#'
+#' plotSpatial(sce, img_id = "ImageNb",
+#'            colPairName = "knn_cn_graph",
+#'            node_color_by = "cellular_neighborhood")
+#'
+#' ## 2. Spatial context (SC)
+#' sce <- buildSpatialGraph(sce, img_id = "ImageNb",
+#'                         type = "knn",
+#'                         name = "knn_sc_graph",
+#'                         k = 15)
+#'
+#' sce <- aggregateNeighbors(sce, colPairName = "knn_sc_graph",
+#'                          aggregate_by = "metadata",
+#'                          count_by = "cellular_neighborhood",
+#'                          name = "aggregatedNeighborhood")
+#'
+#' # Detect spatial context
+#' sce <- detectSpatialContext(sce, entry = "aggregatedNeighborhood",
+#'                            threshold = 0.9)
+#'
+#' plotSpatial(sce, img_id = "ImageNb",
+#'            colPairName = "knn_sc_graph",
+#'            node_color_by = "spatial_context")
+#'            
+#' @seealso 
+#' \code{\link[imcRtools]{filterSpatialContext}} for the function to filter 
+#' spatial contexts
+#'
+#' \code{\link[imcRtools]{plotSpatialContext}} for the function to plot 
+#' spatial context graphs 
+#' 
+#' @author Lasse Meyer (\email{lasse.meyer@@uzh.ch})
+#' 
+#' @references
+#' \href{https://doi.org/10.1016/j.cels.2021.09.012}{
+#' Bhate S. et al., Tissue schematics map the specialization of immune tissue 
+#' motifs and their appropriation by tumors, Cell Systems, 2022}
+#' 
+#' @importFrom SingleCellExperiment colData
+#' @export
+
+detectSpatialContext <- function(object,
+                                 entry = "aggregatedNeighbors",
+                                 threshold = 0.9,
+                                 name = "spatial_context"){
+  
+  .valid.detectSpatialContext.input(object, entry, threshold, name) 
+  
+  cur_dat <- colData(object)[,entry]
+  
+  out_dat <- apply(cur_dat, 1, function(x){
+    
+    out <- cumsum(sort(x, decreasing = TRUE))
+    
+    if (sum(out) != 0) {
+      cur_out <- names(out[seq_len(sum(out < threshold) + 1)])
+      cur_out <- paste(sort(as.numeric(cur_out)), collapse = "_")
```

Comment:
Good point! 

Three main options that I could think of:
1. Add a **validity check** that checks that `as.numeric(colnames(colData(object)[,entry]))` should not give any NA values **OR** 
2. If NA values are produced, change `colnames(cur_dat) <- as.character(seq_along(colnames(cur_dat)))` - then would probably notify the user of that? 
3. I could use something akin to `cur_out <- paste(sort(as.numeric(factor(cur_out))), collapse = "_")` - This would allow conversion of character to numeric, but this way it gives the wrong result due to sorting. 

Probably option 2 is best? Let me know what you think :)